samples/application_development/external_lib/CMakeLists.txt

```diff
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
 cmake_minimum_required(VERSION 3.13.1)
-include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
+find_package(Zephyr 2.2.0 HINTS $ENV{ZEPHYR_BASE})
```

Comment:
To avoid search-replacing these on every Zephyr release I think we should leave the version unset.

Reply:
Thanks for raising this.
I see benefits both in using the version number, and reasons not to do.
So I have no strong preference for one or the other, and I actually believe this could be something to be decided by a wider audience (TSC maybe).

But let me use the chance to describe this a little deeper.

Using `find_package(Zephyr 2.2.0)` will not require us to update on each new revision.
If a Zephyr 2.4.0 is installed, that will also be accepted (with current check mechanism).

also Zephyr 2.6, 2.8, 2.x, 3.x, 4.x and so will all pass.

However, with `find_package(Zephyr 2.2.0 EXACT)`, then we would need to update the version number on each release (agreed, this is not desired).

So what does placing a minimal version number here give as benefit, as it appears to have no meaning.
1) It allows new apps, or apps using a specific feature by default, to state that it requires as minimum Zephyr 2.4.0
As example, `hello_world` could still just require `2.2.0` or above, whereas `new_app` requires `2.4.0` cause it uses a new subsystem in Zephyr.
Thus if someone tries to build an app stating:
`find_package(Zephyr 2.4.0)` but Zephyr is 2.2.0, then they get an error (Out-of-tree builds most likely).
As example:
```
 Could not find a configuration file for package "Zephyr" that is compatible
  with requested version "2.4.0".

  The following configuration files were considered but not accepted:

    C:/projects/ncs/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake, version: 2.2.0
```
2) It allows a major roll on Zephyr, e.g. Zephyr 3.0.0 to request a roll in version number, ensuring a `2.x` app to be updated. 
This is not implemented today, as it must decided by is Zephyr 3.0.0 if it can be used for a 2.x based app.
Introducing such a limitation and apply a check should be done here:
https://github.com/zephyrproject-rtos/zephyr/blob/10478c892f21adbed5635d16ebc66627d1177f5e/share/zephyr-package/cmake/ZephyrConfigVersion.cmake#L11-L27

so today doing `find_package(Zephyr 2.2.0)` is basically equivalent to `find_package(Zephyr)` but where `find_package(Zephyr)` will not allow to check the version in future, then `find_package(Zephyr 2.2.0)` give us the freedom to introduce restrictions if need be.

As long as the referenced code in the link is not changed, then ``find_package(Zephyr 2.2.0)` does never need to be updated.

Reply:
My concern is that users will see the Zephyr version in the app and make one these two incorrect assumptions:

> The Zephyr version I am currently using is 2.2.0.
> 
> This app uses features introduced in 2.2.0 and is therefore not compatible with 2.2.1.

Also, in the future, assuming we don't search-replace on every release, the user might assume that the app works with 2.2.0, when in reality it relies on a feature in 2.3.

I see the value in version checking when needed, but I don't think we should do it for all apps.

But this is just my opinion, if you disagree then I won't NACK over this.

Reply:
If you have no strong preference one way or the other then I would propose dropping the version checking now, and then try to introduce it at a later time.

Then the consequences of merging this PR are less severe and therefore it is easier to merge this PR.

Reply:
> If you have no strong preference one way or the other then I would propose dropping the version checking now, and then try to introduce it at a later time.

+1 Just reading the code it makes an impression that this is the only version supported.

Reply:
> If you have no strong preference one way or the other then I would propose dropping the version checking now, and then try to introduce it at a later time.
> 

The only preference I have, is that this is a decision made by a wider audience.
The explanation was to give a better understanding on what using a version number would imply, and which were the reason for choosing this in the PR (I also considered your suggestion).

But before settling on a preference, let's just list some possibilities:
a) `find_package(Zephyr 2.2.0)` - current suggestion in PR.
b) `find_package(Zephyr 2.2.0 EXACT)` - will require update to ALL samples when version is changed
c) `find_package(Zephyr 2)` or  `find_package(Zephyr 2.)` - make it clearer that any Zephyr 2.x should work
d) `find_package(Zephyr 2.0.0)` - mix between a) and c)
e) `find_package(Zephyr)` - will have limitations described, and downstream users may not be aware that they can use Zephyr version check in their app for increased check possibilities

I personally prefer a) or e) and would advice against b).

Reply:
a) find_package(Zephyr 2.2.0) - current suggestion in PR.
👍 if you prefer this solution (you are allowed to have multiple preferences)

Reply:
b) find_package(Zephyr 2.2.0 EXACT) - will require update to ALL samples when version is changed
👍 if you prefer this solution (you are allowed to have multiple preferences

Reply:
c) find_package(Zephyr 2) or find_package(Zephyr 2.) - make it clearer that any Zephyr 2.x should work
👍 if you prefer this solution (you are allowed to have multiple preferences)

Reply:
d) find_package(Zephyr 2.0.0) - mix between a) and c)
👍 if you prefer this solution (you are allowed to have multiple preferences)

Reply:
e) find_package(Zephyr) - will have limitations described, and downstream users may not be aware that they can use Zephyr version check in their app for increased check possibilities
👍 if you prefer this solution (you are allowed to have multiple preferences)

Reply:
The fact that we are able to provide version information to `find_package` is a very interesting feature, however I wouldn't use any of the options other than e) unless we upgrade the sanitycheck to always run tests with the latest version and the one specified with the `find_package`. Without testing this information is going to degrade very quickly.

Reply:
@mnkp Thanks for the comment, and the vote.

> Without testing this information is going to degrade very quickly.

yes, that is a potential risk.
However, as long as a sample is not modified, that is it's prj.conf and it's sources are not changed, then the sample with default conf should of course work with the specified version.
A simple sanity check could then be, that in case ANY change inside a sample folder, as example inside `hello world` is committed. then sanity check will fail, if Zephyr version specified is not != to current version.

But agreed, keeping the Version out, poses no risk.

Reply:
> A simple sanity check could then be, that in case ANY change inside a sample folder, as example inside `hello world` is committed. then sanity check will fail, if Zephyr version specified is not != to current version.

@tejlmand But in this case we're also going to lie to the user; it may well be the case that the sample application would still work well with an older version of Zephyr. As I mentioned, the ability to provide version information to the `find_package` is an interesting feature and you explained its usage very well in the updated documentation but it will cause more problems than bring benefits if we keep it in the in-tree sample and test applications.

Reply:
Please remove the Zephyr version in the find_package() directive in all samples, there is no way we will ever be able to keep that updated

Reply:
@mnkp @SebastianBoe @finikorg @carlescufi Version number `2.2.0` removed, as it seems that is what the majority prefers.